---
title: "Simulating Data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Simulating Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(simGWAS)
library(DiagrammeR)
library(dplyr)
library(reshape2)
library(ggplot2)
set.seed(1)
```

## Introduction

This vignette demonstrates how to use the `sim_mv` function to simulate data a few different types of GWAS data.

## Introduction to `sim_mv`

The `sim_mv` function generates GWAS summary statistics for multiple continuous traits from a linear structural 
equation model encoded as a matrix of direct effects. Variants can be generated with or without LD.
There are also some helper functions for LD-pruning and generating special kinds of direct effect matrices. 

The `sim_mv` function is a wrapper to a more general function `sim_sumstats_lf` which generates summary statistics given 
a latent factor structure for a set of traits. Direct usage of `sim_sumstats_lf` is demonstrated briefly at the end of this vignette.



## Basic Usage

### Input 

The `sim_mv` function has five required arguments:

+ `N`: The GWAS sample size for each trait. This can be a scalar, vector or matrix. If a vector, `N` should have length equal to the number of traits. If there are overlapping samples between GWAS, `N` should be a matrix (see "Sample Overlap" below). 
+ `J`: The number of SNPs to simulate (scalar).
+ `h2`: The hertiability of each trait. This can be a scalar or a vector with length equal to the number of traits generated.
+ `pi`: The proportion of all SNPs that have a direct effect on each trait. This can be a scalar or a vector with length equal to the number of traits generated.
+ `G`: A matrix specifying direct effects in the linear structural equation model. 

There are additional optional arguments:

+ `R_E`: Matrix giving correlation between environmental components of each trait (see "Sample Overlap" for more details). Defaults to the identity. `R_E` only affects summary statistics when there is sample overlap.
+ `R_LD` and `snp_info`: Options used if GWAS data should have LD (See "Simulating Data with LD"). 
+ `af`: Optional vector of allele frequencies. This option is only used if data have no LD.
+ `sporadic_pleiotropy`: Allow a single variant to have direct effects on multiple traits. Defaults to TRUE.
+ `return_dat`: A developer option, return some extra data that is useful for debugging and testing. 

### Output

The `sim_mv` function returns a list with the following elements

+ `beta_hat`, `se_beta_hat`: Simulated GWAS effect estimates and standard errors
+ `beta_joint`: Total causal effects of SNPs on traits. 
+ `beta_marg`: Expected marginal association of SNPs on traits. `beta_marg` is the expected value of `beta_hat`. When there is no 
LD, `beta_marg` and `beta_joint` are the same. 
+ `direct_SNP_effects_joint`: Direct causal effects of SNPs on traits. Direct means not mediated by other traits. 
+ `direct_SNP_effects_marg`: Like `beta_marg` but considering only direct rather than total effects. 
+ `trait_corr`: Population trait correlation. This is determined by genetic sharing (from trait heritability and `G`) and environmental correlation, `R_E`.
+ `R`: Correlation in sampling error of `beta_hat` across traits. This is the identity if there is no
sample overlap. Otherwise it is a function of `trait_corr` and the amount of sample overlap. 
+ `direct_trait_effects`: Matrix of direct effects between traits
+ `total_trait_effects`: Matrix of total effects between traits

The order of the columns in `beta_hat`, `se_beta_hat`, `direct_SNP_effects_joint`, 
`direct_SNP_effects_marg`, `beta_joint`, and `beta_marg` corresponds
to the order of variables in `G`. 

## Specifying the DAG

The matrix `G` specifies a linear structural equation model for a set of traits. `G` must be a square matrix with dimension $K\times K$ where $K$ is the total number of traits. That is, `G` must contain a row and column for every trait even if that trait is independent of all other traits. 
The `G[i,j]` entry of `G` specifies the direct linear effect of trait $i$ on trait $j$. The
diagonal entries of $G$ should be 0. An error will be generated if `G` specifies a graph that contains cycles. All traits are assumed to have variance 
equal to 1, so `G[i,j]^2` is the proportion of trait $j$ variance explained by 
the direct effect of trait $i$. 

For example, the graph below: 

```{r, echo=FALSE, fig.align='center', fig.width = 5}
G <- matrix(c(0, sqrt(0.25), 0, sqrt(0.15), 
              0, 0, 0, sqrt(0.1), 
              sqrt(0.2), 0, 0, -sqrt(0.3), 
              0, 0, 0, 0), nrow = 4, byrow = TRUE)


d <- melt(G) %>%
     filter(value !=0) %>%
     rename(from = Var1, to = Var2)


n <- create_node_df(n = 4, label = c("X", "Y", "Z", "W"), 
                    fontname = "Helvetica", 
                    fontsize = 10, 
                    width = 0.3, 
                    fillcolor = "white", 
                    fontcolor = "black",
                    color = "black", 
                    x = c(0, 1, 1, 2), 
                    y = c(0, -0.5, 1, 0))
e <- create_edge_df(from = d$from, to = d$to, minlen = 1,  color = "black", 
                    label = round(d$value, digits = 3))
g <- create_graph(nodes_df = n, edges_df = e)

render_graph(g)
```

is represented by the matrix

```{r}
G <- matrix(c(0, sqrt(0.25), 0, sqrt(0.15), 
              0, 0, 0, sqrt(0.1), 
              sqrt(0.2), 0, 0, -sqrt(0.3), 
              0, 0, 0, 0), nrow = 4, byrow = TRUE)
colnames(G) <- row.names(G) <- c("X", "Y", "Z", "W")
G
```

To simulate data from this graph, we can use

```{r}
sim_dat1 <- sim_mv(G = G,
                  N = 60000, J = 50000, 
                  h2 = c(0.3, 0.3, 0.5, 0.4), 
                  pi = 1000/50000)
```

In the function call above, we specified a GWAS sample size of 60k for every GWAS with no overlap. 
We gave different heritabilities of each of the four traits (the `h2` argument). 
Every trait has an expected 1000 direct effect SNPs (the `pi` argument). Note that `pi` specifies
the expected proportion of direct effect SNPs not the expected number of direct effect SNPs.
`J` specifies the total number of variants to generate, in this case 50,000.


### Generating $G$ from "XYZ" mode.

The function `xyz_to_G` will generate a matrix, $G$, corresponding to a specific "XYZ" DAG form. 
In the "XYZ" DAG, 
there is an exposure ($X$), an outcome ($Y$), and $K-2$ other variables, $Z_1, \dots, Z_{K-2}$. 

There is a (possibly 0) effect of $X$ on $Y$ specified by the `gamma` argument. 
Variables $Z_1, \dots, Z_{K-2}$ can have effects either on or from $X$ and/or $Y$ or both
but there are no affects $Z_1, \dots, Z_{K-2}$ on each other. 
Effects between each $Z_k$ and $X$ and $Y$ respectively are given in the `tau_xz` and `tau_yz` arguments. 
The direction of these effects is given in the `dir_xz` and `dir_yz` arguments. All four of
these arguments should have length $k-2$. 

The direction parameters `dir_xz` and `dir_yz` are vectors of 1 or -1 with 1 indicating an effect on $X$ or $Y$ and -1 indicating an effect from $X$ or $Y$. 
Effect size arguments `gamma`, `tau_xz`, and `tau_yz` are given as signed proportion of variance explained. 
So if `gamma = -0.3`, The direct effect of $X$ explains 30\% of the variance of $Y$ and the 
effect of $X$ on $Y$ is negative. 






For example, the code

```{r}
myG <- xyz_to_G(tau_xz = c(0.1, -0.15, 0.2, 0.3), 
                   tau_yz = c(0, 0.2, -0.25, 0.15), 
                   dir_xz = c(1, 1, -1, -1), 
                   dir_yz = c(1, 1, -1, 1),
                   gamma = 0.3)
```

generates the matrix corresponding to the graph below:


```{r, echo=FALSE, fig.align='center', fig.width = 5}

d <- melt(myG) %>%
     filter(value !=0) %>%
     rename(from = Var1, to = Var2)

n <- create_node_df(n = 6, label = c("Y", "X", "Z1", "Z2", "Z3", "Z4"), 
                    fontname = "Helvetica", 
                    fontsize = 10, 
                    width = 0.3, 
                    fillcolor = "white", 
                    fontcolor = "black",
                    color = "black", 
                    x = c(2, 0, -0.5, 1, 1, 1), 
                    y = c(0, 0, 1, 1, -0.5, -1))
e <- create_edge_df(from = d$from, to = d$to, minlen = 1,  color = "black", 
                    label = round(d$value, digits = 3))
g <- create_graph(nodes_df = n, edges_df = e)

render_graph(g)
```


The weights in the graph give the effect size. Note that this is the square root of the value provided in `tau_xz` and `tau_yz` which specifies the percent variance explained. For example, the effect of $Z_1$ on $X$ is  $0.316 = \sqrt{0.1}$ and the effect of $Z_2$ on $X$ is $-0.387 = - \sqrt{0.15}$. 

## A Closer Look at the Output 

We can now take a look at the output. Summary statistics are contained in the `beta_hat` and `se_beta_hat` tables which both have dimension $J\times K$.
```{r}
names(sim_dat1)
dim(sim_dat1$beta_hat)
head(sim_dat1$beta_hat)
head(sim_dat1$se_beta_hat)
```

The `direct_trait_effects` object is a matrix giving the input `G` while `total_trait_effects` gives the total effect of each trait on each other trait. 
```{r}
sim_dat1$direct_trait_effects
sim_dat1$total_trait_effects
```

True variant effects are also stored in the simulation object. The `direct_SNP_effects_joint` and `beta_joint` objects give the direct and total SNP effects. The `_joint` ending here indicates that these variables store the expected *joint* association, conditional on all other variants, i.e. the causal effects. GWAS effect estimates measure the marginal association, so we also store the expected direct and total *marginal* associations in 
`direct_SNP_effects_marg` and `beta_marg`. When all variants are independent (no LD),  the expected joint and marginal associations are the same. 

```{r}
all.equal(sim_dat1$beta_joint, sim_dat1$beta_marg)
```

However, if we had generated data with LD, this would not be the case (see below for more on LD). `beta_marg` is always the expected value of `beta_hat`, that is `beta_marg` is the association we would measure with infinite GWAS sample size. 

Direct SNP effects are always independent across traits while total SNP effects are the sum of direct effects and indirect effects mediated by other traits. To see the difference, we make some plots.

First we plot direct SNP effects on $Z$ vs direct SNP effects on $W$
```{r}
plot(sim_dat1$direct_SNP_effects_joint[,3], sim_dat1$direct_SNP_effects_joint[,4], 
     xlab = "Direct Z effect", ylab = "Direct W effect")
```

Because direct effects are generated from a point-normal distribution, most variants have direct effects on at most one of $Z$ or $W$. However, there are a handful of variants that have direct effects on both. This is "sporadic pleiotropy" and can be prevented with `sporadic_pleiotropy=FALSE` if it is not desired. 

Next we plot the total SNP effects on $Z$ vs the total SNP effects on $W$. Because $Z$ has a causal effect on $W$, all variants with effects on $Z$ also affect $W$. The line in the plot has slope equal to the total effect of $Z$ on $W$. The majority SNPs that have non-zero effect on $Z$ fall exactly on this line. With `sporadic_pleiotropy= FALSE`, all of the variants with non-zero effect on $Z$ would fall on this line. The variants on the vertical line at 0 are variants with non-zero direct effect on $W$ but no direct effect on $Z$. 

```{r}
plot(sim_dat1$beta_joint[,3], sim_dat1$beta_joint[,4], 
     xlab = "Total Z effect", ylab = "Total W effect")
abline(0, sim_dat1$total_trait_effects[3,4], col = "red", lty = 2, lwd = 2)
```

Finally, we can verify that the observed GWAS effect estimate is a noisy measurement of `beta_marg`.

```{r}
plot(sim_dat1$beta_marg[,3], sim_dat1$beta_hat[,3], 
     xlab = "beta_marg[,3] = E[beta_hat[,3]]", ylab = "beta_hat[,3]")
abline(0, 1, col = "red", lty = 2, lwd = 2)
```

## Standardized vs Non-Standardized Effects

By default, when there is no LD, `sim_mv` returns standardized effects and standardized effect estimates. These are the estimates we would obtain if genotypes were scaled to have population variance 1 rather than being coded as 0, 1, and 2. The relationship between standardized and non-standardized effect estimates is simply $\hat{\beta}_j^{(non-std)} = \hat{\beta}_j^{(std)}/sd(g_j)$ where $j$ indexes variants, $sd(g_j)$ is the population standard deviation of variant $j$ (i.e. $\sqrt{2 f_j (1-f_j)}$ where $f_j$ is the allele frequency of variant $j$). Similarly the relationship between the standard error of standardized and non-standardized effect estimates is $se\left(\hat{\beta}_j^{(non-std)}\right) = se\left(\hat{\beta}_j^{(std)}\right)/sd(g_j)$. 

In our previously generated `sim_dat1`, every effect estimate has the same standard error. This is because the standard error of $\hat{\beta}_j^{(std)}$ is approximated as $1/\sqrt{N}$ and we used a common sample size across all traits. Non-standardized effect estimates will have different standard errors for different variants because these also depend on variant allele frequency. 


Generating non-standardized effects requires input allele frequencies which are used to calculate the variance of each SNP. This can be provided by the `af` argument which can accept either a scalar, in which case the same allele frequency is used for every SNP, a vector of length `J`, or a function that takes a single argument and returns a vector of allele frequencies with length determined by the argument. If the `af` argument is provided, `sim_mv` will return both effect estimates and expected associations/true effects on the non-standardized scale.


```{r}
sim_dat2 <- sim_mv(G = G,
                  N = 60000, J = 50000, 
                  h2 = c(0.3, 0.3, 0.5, 0.4), 
                  pi = 1000/50000, 
                  af = function(n){rbeta(n, 1, 5)})

head(sim_dat2$se_beta_hat)
```

## Simulating Data with LD

`sim_mv` can be used to generate data with LD by inputting a list of LD matrices and a corresponding table of SNP information. The function will work fastest if the LD matrix is broken into small blocks. The input data format for the LD pattern is a list of either a) matrices, b) sparse matrices (class `dsCMatrix`) or c) eigen decompositions (class `eigen`). The `simGWAS` package contains a built in data set containing the LD pattern from Chromosome 19 in HapMap3 broken into 39 blocks. This LD was estimated from the HapMap3 European subset using LDShrink. This data set can also be downloaded [here](https://zenodo.org/record/6761943#.Yrno2njMIUE). The LD pattern must be accompanied by a data frame containing columns `SNP` giving a SNP name (can be arbitrary) and `AF` giving the allele frequency.

Let's look at the built-in LD data

```{r}
data("ld_mat_list")
data("snpdata")

length(ld_mat_list)

sapply(ld_mat_list, class)
```




```{r}
# This prints the number of SNPs in each block
sapply(ld_mat_list, nrow)

sapply(ld_mat_list, nrow) %>% sum()

dim(snpdata)

head(snpdata)
```

The LD pattern covers 19,490 SNPs. The number of rows in the variant annotation data frame must match the number of variants in the LD pattern. If we wish to generate GWAS data for a larger number of variants than are covered by the LD pattern, `simGWAS` will 
repeat the LD pattern as many times as necessary to give the desired number of variants. 


To generate data with LD, the LD pattern and the variant annotation data frame are given to the `R_LD` and `snp_info` parameters.  The built-in LD pattern corresponds to a density of about 1.2 million variants in per genome. However, in the interest of saving time and space, we generate data for only 100k variants. This means that causal effects will be denser than they might be in more realistic data.

```{r}
set.seed(10)
sim_dat1_LD <- sim_mv(G = G,
                  N = 60000, J = 1e5, 
                  h2 = c(0.3, 0.3, 0.5, 0.4), 
                  pi = 1000/1e5, R_LD = ld_mat_list, 
                  snp_info = snpdata)

```

In data with LD, the `_joint` objects and `_marg` objects are not identical. For example, we can compare `beta_joint` and `beta_marg` for $Z$. 

```{r}
with(sim_dat1_LD, plot(beta_joint[,3], beta_marg[,3]))
abline(0, 1, lty = 2, lwd =2, col = "red")
```

We can consider variants with non-zero values of `beta_joint[,3]` to be variants with causal effects on $Z$ while those with non-zero values of `beta_marg[,3]` have non-zero (population) marginal association with $Z$, meaning that they are in LD with at least one causal variant. In the plot, we see that many variants with no causal effect nevertheless have non-zero marginal association, as expected. The causal variants don't fall exactly on the red line because in some cases, multiple causal variants are in LD with each other. 

### LD-Pruning, LD-proxies, and LD matrix extraction

Many post-GWAS applications such as Mendelian randomization and polygenic risk score construction require an LD-pruned set of variants. `simGWAS` contains a few LD-related functions to help with pruning and testing methods that require input LD matrices. Note that all of these methods use the true LD pattern rather than estimated LD. 

The `sim_ld_prune` function will perform LD-clumping on simulated data, prioritizing variants according to a supplied `pvalue` vector. Although this argument is called `pvalue`, it can be any numeric vector used to prioritize variants. The `pvalue` argument can also accept an integer. If `pvalue = i`,  variants will be prioritized according to the p-value for the trait in the column `i`. If `pvalue` is omitted, variants will be prioritized randomly (so a different result will be obtained each re-run unless a seed  is set). 

To speed up performance, if you only need variants with $p$-value less than a certain threshold, supply the `pvalue_thresh` argument. Below we prune based on the p-values for trait $Z$ in two different equivalent ways.

```{r}
pruned_set1 <- sim_ld_prune(dat = sim_dat1_LD, 
                            pvalue = 3, 
                            R_LD = ld_mat_list, 
                            r2_thresh = 0.1,
                            pval_thresh = 1e-6)
length(pruned_set1)
pval3 <- with(sim_dat1_LD, 2*pnorm(-abs(beta_hat[,3]/se_beta_hat[,3])))
pruned_set2 <- sim_ld_prune(dat = sim_dat1_LD, 
                            pvalue = pval3, 
                            R_LD = ld_mat_list, 
                            r2_thresh = 0.1,
                            pval_thresh = 1e-6)
all.equal(pruned_set1, pruned_set2)
```
`sim_ld_prune` returns a list of indices corresponding to an LD-pruned set of variants. 

The `sim_ld_proxy` function will return LD-proxies (variants with LD above a given threshold) with a supplied set of variants. Here we extract proxies for a few arbitrary variants. The `return_mat` option will cause the function to return the LD matrix for the proxies as well as the indices of proxies

```{r}
ld_proxies <- sim_ld_proxy(sim_dat1_LD, index = c(100, 400, 600), R_LD = ld_mat_list, r2_thresh = 0.64, return_mat = TRUE)
ld_proxies
```

Finally, the `sim_extract_ld` function will extract the LD matrix for a set of variants. We can use this to check the previous functions. First, both of the proxies for index 600 should have correlation greater than 0.8 with index 600. 

```{r}
ld_mat1 <- sim_extract_ld(sim_dat1_LD, index = c(600, ld_proxies[[3]]$proxy_index), R_LD = ld_mat_list)
ld_mat1
all(ld_mat1[,1] > 0.8)
```

Next we can check that the LD-pruned set have no mutual correlations greater than $\sqrt(0.1) \approx 0.316$

```{r}
ld_mat2 <- sim_extract_ld(sim_dat1_LD, index = pruned_set1, R_LD = ld_mat_list)
diag(ld_mat2) <- 0 # remove the diagonal
all(ld_mat2^2 < 0.1)
```

## Sample Overlap

If two GWAS are performed on different traits using overlapping samples, the sampling errors of effect estimates will be correlated.
If the two GWAS have sample sizes $N_1$ and $N_2$ with $N_c$ overlapping samples, then the correlation of 
$\hat{z}_{1j}$ and $\hat{z}_{2j}$, $z$-scores for variant $j$ in study 1 and study 2, is approximately $\frac{N_c}{\sqrt{N_1 N_2}} \rho_{1,2}$ where $\rho_{1,2}$ is the trait correlation. 

In `sim_mv` it is possible to specify sample overlap between studies by providing `N` as a matrix rather than a scalar or a vector. If `N` is a matrix than `N[i,i]` is the sample size of study $i$ and `N[i,j]` is the number of overlapping samples between studies $i$ and $j$. In the model used by `simGWAS`, all traits can be decomposed into a genetic component and an environmental component. We always assume the 
environmental trait component is independent of the genetic trait component. Correlation between the genetic components of the specified traits is determined by the supplied DAG. It is also possible to supply the correlation between the environmental components of the traits using the `R_E` parameter. 


Below is an example using the same 4 trait DAG we have been using so far. Below we specify that all four GWAS have sample size 60,000 and 
there are 30,000 samples overlapping between any pair of GWAS.
Note that `simGWAS` assumes that the trait correlation is the same in any subset of samples so it doesn't matter if the same 30,000
samples are in all studies or if different sets overlap between different pairs of studies.
We also specify an environmental correlation matrix in which the pairwise correlation in environmental effects for any pair of traits is 0.4. `R_E` defaults to the identity if it is not provided.

```{r}
R_E <- matrix(0.4, nrow = 4, ncol = 4)
diag(R_E) <- 1
N <- matrix(30000, nrow = 4, ncol = 4)
diag(N) <- 60000
N
```

```{r}
set.seed(10)
sim_dat3 <- sim_mv(G = G,
                  N = N, J = 1e5, R_E = R_E,
                  h2 = c(0.3, 0.3, 0.5, 0.4), 
                  pi = 1000/1e5)
```

The simulation data object contains three matrices that describe between study correlation. `R_E` is the environmental correlation supplied in the command option, `trait_corr` is the total trait correlation, and `R` is the correlation between noise terms for pairs of GWAS which is obtained by scaling `trait_corr` by the appropriate values.

```{r}
sim_dat3$R_E
sim_dat3$trait_corr
sim_dat3$R
```

If there is no sample overlap, the `R_E` argument to `sim_mv` will have no affect on summary statistics but will influence the calculation of `trait_corr`. For example, in our previous data generated without sample overlap and in which we did not specify `R_E` we have
```{r}
sim_dat1_LD$R_E
sim_dat1_LD$trait_corr
sim_dat1_LD$R
```

The `R` matrix will always be the identity if there is no sample overlap. 

## Generating Data for Multiple GWAS with the Same Variant Effects

In some cases, it may be desirable to generate multiple sets of summary statistics for the same set of effect sizes. This
would mimic performing multiple GWAS on the same trait. The function `gen_bhat_from_b` allows this. This function takes as input
a matrix of either standardized or non-standardized joint (causal) effects to `b_joint_std` or `b_joint`. The function also requires the trait correlation matrix
and the sample size argument. Optionally, LD arguments `R_LD` and `snp_info` can be supplied, or the allele frequency can be supplied via `af`. Most users will not need the other options.

Below, we generate new GWAS data for the effects in `sim_dat1_LD`. Here we use the same LD pattern used originally but this is not required. For example, we could simulate GWAS in samples with different ancestry but supplying a different LD pattern.

```{r}
sim_dat2_LD <- gen_bhat_from_b(b_joint = sim_dat1_LD$beta_joint, 
                               trait_corr = sim_dat1_LD$trait_corr, 
                               N = 40000, 
                               R_LD = ld_mat_list, snp_info = snpdata)
```
The resulting simulation data object only contains effect estimates, standard errors, `R`, and `snp_info`. All other information should match the original data set. 

Below we compare effect estimates in the new study and the old study for trait $X$.
```{r}
plot(sim_dat2_LD$beta_hat[,1], sim_dat1_LD$beta_hat[,1])
abline(0, 1, col = "red", lty = 2, lwd = 2)
```

Our first simulation object `sim_dat1` was generated with no LD and without supplying the `af` argument, so the `beta_joint` values for that object represent standardized effects. If we were generating secondary studies from this object, we should use the `b_joint_std` argument rather than `b_joint` in the call to `get_bhat_from_b`.
