<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="simGWAS">
<title>Controlling Effect Size Distributions • simGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Controlling Effect Size Distributions">
<meta property="og:description" content="simGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">simGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0.0068</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/effect_distribution.html">Controlling Effect Size Distributions</a>
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/simGWAS/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Controlling Effect Size Distributions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jean997/simGWAS/blob/HEAD/vignettes/effect_distribution.Rmd" class="external-link"><code>vignettes/effect_distribution.Rmd</code></a></small>
      <div class="d-none name"><code>effect_distribution.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jean997/simGWAS" class="external-link">simGWAS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>simGWAS</code> can support a range of effect size
distributions. At one end of the spectrum, effect sizes and direct
effect variants can be completely deterministic. At the other end of the
spectrum, both are random. There are multiple ways to control the
distributions of these random values demonstrated in this vigentte.</p>
</div>
<div class="section level2">
<h2 id="default-behavior">Default Behavior<a class="anchor" aria-label="anchor" href="#default-behavior"></a>
</h2>
<p>The default behavior of <code>sim_mv</code> is to sample effect
variants randomly by first generating a <span class="math inline">\(J\)</span>-vector <span class="math inline">\(Z_m
\sim Binom(1, \pi_m)\)</span> for each trait <span class="math inline">\(m\)</span>. <span class="math inline">\(Z_{m,j}\)</span> is 1 if variant <span class="math inline">\(j\)</span> has a direct effect on trait <span class="math inline">\(m\)</span> and 0 otherwise. We then sample
standardized effect sizes for variants with direct effects from a normal
<span class="math inline">\(N(0, \sigma^2_m)\)</span> distribution with
<span class="math inline">\(\sigma^2_m\)</span> determined by the SEM
and the target expected heritability. This results in effect variants
spread evenly throughout the genome and effect sizes that are inversely
proportional to <span class="math inline">\(\sqrt{2 f_j
(1-f_j)}\)</span></p>
<p>We demonstrate this by simulating data for only one trait.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">sim_dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-3-2.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">std_effects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat1</span>, <span class="va">direct_SNP_effects_joint</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">std_effects</span><span class="op">[</span><span class="op">!</span><span class="va">std_effects</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, breaks <span class="op">=</span> <span class="fl">30</span>, freq <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"Standardized Direct Effects"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/curve.html" class="external-link">curve</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">dnorm</a></span><span class="op">(</span><span class="va">x</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.3</span><span class="op">/</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>, add <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-3-3.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="controlling-direct-effect-variants">Controlling Direct Effect Variants<a class="anchor" aria-label="anchor" href="#controlling-direct-effect-variants"></a>
</h2>
<p>To control which variants have direct effects, the <code>pi</code>
argument can be supplied as a <code>J \times M</code> matrix rather than
as a vector of length <code>M</code> or a scalar. If <code>pi</code> has
matrix format then <code>pi[j,m]</code> gives the probability than
variant <code>j</code> has a direct effect on trait <code>m</code>. To
make the set of direct effect variants fully deterministic, set the
values of <code>pi</code> to be either 0 or 1. Using <code>pi</code> in
matrix format is incompatible with options <code>pi_exact = TRUE</code>
and <code>sporadic_pleiotropy = FALSE</code>.</p>
<p>Below, we use the matrix format to force all effect variants to be in
the first half of the genome. This strategy could be used to designate
regions with high concentrations of effect variants.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pi_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fl">50000</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pi_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25000</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">25000</span></span>
<span><span class="va">sim_dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="va">pi_mat</span>, </span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>In another application of this option, we can make the probability
that a variant is an effect variant a function of the allele
frequency.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">50000</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">pi_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fl">50000</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pi_mat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">af</span>, <span class="fl">1</span><span class="op">-</span><span class="va">af</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">4</span></span>
<span><span class="co"># ensure there are still 1000 effect variants expected</span></span>
<span><span class="va">pi_mat</span> <span class="op">&lt;-</span> <span class="fl">1000</span><span class="op">*</span><span class="va">pi_mat</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pi_mat</span><span class="op">)</span> </span>
<span><span class="va">sim_dat3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="va">pi_mat</span>, </span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<p>To verify that we observe the desired effect, we plot the proportion
of effect variants by bins of minor allele frequency.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.5</span>, length.out <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">af</span>, <span class="fl">1</span><span class="op">-</span><span class="va">af</span><span class="op">)</span></span>
<span><span class="va">snp_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">maf</span>, </span>
<span>                      is_effect <span class="op">=</span> <span class="va">sim_dat3</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span> <span class="op">!=</span><span class="fl">0</span>, </span>
<span>                      maf_cat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">maf</span>, breaks <span class="op">=</span> <span class="va">x</span><span class="op">)</span>, </span>
<span>                      maf_cat_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">maf</span>, breaks <span class="op">=</span> <span class="va">x</span>, labels <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">maf_cat</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>                     n_eff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_effect</span><span class="op">)</span>, </span>
<span>                     maf_min <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/nth.html" class="external-link">first</a></span><span class="op">(</span><span class="va">maf_cat_order</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>eff_prob <span class="op">=</span> <span class="va">n_eff</span><span class="op">/</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">snp_dat</span><span class="op">$</span><span class="va">maf_min</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">snp_dat</span><span class="op">$</span><span class="va">maf_min</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">snp_dat</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">maf_min</span>, <span class="va">eff_prob</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="controlling-effect-size-distribution">Controlling Effect Size Distribution<a class="anchor" aria-label="anchor" href="#controlling-effect-size-distribution"></a>
</h2>
<p>To sample standardized effects from a distribution other than the
normal distribution, we use the <code>snp_effect_function</code>
argument in <code>sim_mv</code>. This argument can accept a function
that accepts three arguments, <code>n</code> for number of variants,
<code>sd</code> for total variance, and <code>af</code> for allele
frequency. The function should return a vector of standardized effect
sizes <span class="math inline">\(\tilde{\gamma}_1, \dots,
\tilde{\gamma}_n\)</span> such that <span class="math inline">\(\sum_{j
= 1}^n E[\tilde{\gamma}_j^2] = \sigma^2\)</span> where <span class="math inline">\(\sigma^2\)</span> is equal to <code>sd^2</code>
and <span class="math inline">\(E[\tilde{\gamma}_j] = 0\)</span>. Below
are two examples using this argument.</p>
<div class="section level3">
<h3 id="drawing-effects-from-a-mixture-of-normals">Drawing Effects from a Mixture of Normals<a class="anchor" aria-label="anchor" href="#drawing-effects-from-a-mixture-of-normals"></a>
</h3>
<p>Frequently, we observe a pattern that many effect variants have a
small “background” effect and some have a larger “forground” effect.
This pattern is consistent with an omni-genic model for genetic
architecture. To simulate this, we can draw standardized effects from a
mixture of normals rather than from a single normal distribution. There
is a helper function <code>mixnorm_to_scale_fam</code> which generates
functions approrpriate to pass to <code>sim_mv</code>. For example,
suppose we want a distribution in which 90% of variants have
standardized direct effects following a <span class="math inline">\(N(0,
\sigma^2_1)\)</span> distribution and 10% have standardized effects from
a <span class="math inline">\(N(0, 10 \sigma^2_1)\)</span> distribution.
Then we can use</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="fu">mixnorm_to_scale_fam</span><span class="op">(</span>sigma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that the <code>sigma</code> argument just specifies the relative
scalings of the variances so it would be equivalent to use</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="fu">mixnorm_to_scale_fam</span><span class="op">(</span>sigma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sim_dat4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="fl">5000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                   snp_effect_function <span class="op">=</span> <span class="va">myf</span>,</span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<p>This gives a distribution of standardized direct effects with much
heavier tails than using the default behavior.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">std_effects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat4</span>, <span class="va">direct_SNP_effects_joint</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">std_effects</span><span class="op">[</span><span class="op">!</span><span class="va">std_effects</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, breaks <span class="op">=</span> <span class="fl">30</span>, freq <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"Standardized Direct Effects"</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Note that even if your function does not use the allele frequency, it
must be able to accept allele frequency as an argument. An easy way to
allow this is to give <code>...</code> as one of the arguments to your
function.</p>
</div>
</div>
<div class="section level2">
<h2 id="allowing-effect-size-to-depend-on-allele-frequency">Allowing Effect Size to Depend on Allele Frequency<a class="anchor" aria-label="anchor" href="#allowing-effect-size-to-depend-on-allele-frequency"></a>
</h2>
<p>We can use the custom effect size function argument to control how
effect size depends on allele frequency. In the LDAK model,</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
