<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="simGWAS">
<title>Simulating Data • simGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating Data">
<meta property="og:description" content="simGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">simGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0.0039</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/simGWAS/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="simulating_data_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><script src="simulating_data_files/viz-1.8.2/viz.js"></script><link href="simulating_data_files/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="simulating_data_files/grViz-binding-1.0.9/grViz.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulating Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jean997/simGWAS/blob/HEAD/vignettes/simulating_data.Rmd" class="external-link"><code>vignettes/simulating_data.Rmd</code></a></small>
      <div class="d-none name"><code>simulating_data.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jean997/simGWAS" class="external-link">simGWAS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rich-iannone/DiagrammeR" class="external-link">DiagrammeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates how to use the <code>sim_mv</code>
function to simulate data a few different types of GWAS data.</p>
</div>
<div class="section level2">
<h2 id="introduction-to-sim_mv">Introduction to <code>sim_mv</code><a class="anchor" aria-label="anchor" href="#introduction-to-sim_mv"></a>
</h2>
<p>The <code>sim_mv</code> function generates GWAS summary statistics
for multiple continuous traits from a linear structural equation model
encoded as a matrix of direct effects. Variants can be generated with or
without LD. There are also some helper functions for LD-pruning and
generating special kinds of direct effect matrices.</p>
<p>The <code>sim_mv</code> function is a wrapper to a more general
function <code>sim_sumstats_lf</code> which generates summary statistics
given a latent factor structure for a set of traits. Direct usage of
<code>sim_sumstats_lf</code> is demonstrated briefly at the end of this
vignette.</p>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<div class="section level3">
<h3 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h3>
<p>The <code>sim_mv</code> function has five required arguments:</p>
<ul>
<li>
<code>N</code>: The GWAS sample size for each trait. This can be a
scalar, vector or matrix. If a vector, <code>N</code> should have length
equal to the number of traits. If there are overlapping samples between
GWAS, <code>N</code> should be a matrix (see “Sample Overlap”
below).</li>
<li>
<code>J</code>: The number of SNPs to simulate (scalar).</li>
<li>
<code>h2</code>: The hertiability of each trait. This can be a
scalar or a vector with length equal to the number of traits
generated.</li>
<li>
<code>pi</code>: The proportion of all SNPs that have a direct
effect on each trait. This can be a scalar or a vector with length equal
to the number of traits generated.</li>
<li>
<code>G</code>: A matrix specifying direct effects in the linear
structural equation model.</li>
</ul>
<p>There are additional optional arguments:</p>
<ul>
<li>
<code>R_E</code>: Matrix giving correlation between environmental
components of each trait (see “Sample Overlap” for more details).
Defaults to the identity. <code>R_E</code> only affects summary
statistics when there is sample overlap.</li>
<li>
<code>R_LD</code> and <code>snp_info</code>: Options used if GWAS
data should have LD (See “Simulating Data with LD”).</li>
<li>
<code>af</code>: Optional vector of allele frequencies. This option
is only used if data have no LD.</li>
<li>
<code>sporadic_pleiotropy</code>: Allow a single variant to have
direct effects on multiple traits. Defaults to TRUE.</li>
<li>
<code>return_dat</code>: A developer option, return some extra data
that is useful for debugging and testing.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The <code>sim_mv</code> function returns a list with the following
elements</p>
<ul>
<li>
<code>beta_hat</code>, <code>se_beta_hat</code>: Simulated GWAS
effect estimates and standard errors</li>
<li>
<code>beta_joint</code>: Total causal effects of SNPs on
traits.</li>
<li>
<code>beta_marg</code>: Expected marginal association of SNPs on
traits. <code>beta_marg</code> is the expected value of
<code>beta_hat</code>. When there is no LD, <code>beta_marg</code> and
<code>beta_joint</code> are the same.</li>
<li>
<code>direct_SNP_effects_joint</code>: Direct causal effects of SNPs
on traits. Direct means not mediated by other traits.</li>
<li>
<code>direct_SNP_effects_marg</code>: Like <code>beta_marg</code>
but considering only direct rather than total effects.</li>
<li>
<code>trait_corr</code>: Population trait correlation. This is
determined by genetic sharing (from trait heritability and
<code>G</code>) and environmental correlation, <code>R_E</code>.</li>
<li>
<code>R</code>: Correlation in sampling error of
<code>beta_hat</code> across traits. This is the identity if there is no
sample overlap. Otherwise it is a function of <code>trait_corr</code>
and the amount of sample overlap.</li>
<li>
<code>direct_trait_effects</code>: Matrix of direct effects between
traits</li>
<li>
<code>total_trait_effects</code>: Matrix of total effects between
traits</li>
</ul>
<p>The order of the columns in <code>beta_hat</code>,
<code>se_beta_hat</code>, <code>direct_SNP_effects_joint</code>,
<code>direct_SNP_effects_marg</code>, <code>beta_joint</code>, and
<code>beta_marg</code> corresponds to the order of variables in
<code>G</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="specifying-the-dag">Specifying the DAG<a class="anchor" aria-label="anchor" href="#specifying-the-dag"></a>
</h2>
<p>The matrix <code>G</code> specifies a linear structural equation
model for a set of traits. <code>G</code> must be a square matrix with
dimension <span class="math inline">\(K\times K\)</span> where <span class="math inline">\(K\)</span> is the total number of traits. That is,
<code>G</code> must contain a row and column for every trait even if
that trait is independent of all other traits. The <code>G[i,j]</code>
entry of <code>G</code> specifies the direct linear effect of trait
<span class="math inline">\(i\)</span> on trait <span class="math inline">\(j\)</span>. The diagonal entries of <span class="math inline">\(G\)</span> should be 0. An error will be generated
if <code>G</code> specifies a graph that contains cycles. All traits are
assumed to have variance equal to 1, so <code>G[i,j]^2</code> is the
proportion of trait <span class="math inline">\(j\)</span> variance
explained by the direct effect of trait <span class="math inline">\(i\)</span>.</p>
<p>For example, the graph below:</p>
<div id="htmlwidget-0dd1ba32e0541469b5ed" style="width:480px;height:432.632880098888px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-0dd1ba32e0541469b5ed">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"X\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"0,0!\"] \n  \"2\" [label = \"Y\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"1,-0.5!\"] \n  \"3\" [label = \"Z\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"1,1!\"] \n  \"4\" [label = \"W\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"2,0!\"] \n\"3\"->\"1\" [color = \"black\", label = \"0.447\"] \n\"1\"->\"2\" [color = \"black\", label = \"0.5\"] \n\"1\"->\"4\" [color = \"black\", label = \"0.387\"] \n\"2\"->\"4\" [color = \"black\", label = \"0.316\"] \n\"3\"->\"4\" [color = \"black\", label = \"-0.548\"] \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p>is represented by the matrix</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.25</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.15</span><span class="op">)</span>, </span>
<span>              <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.2</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.3</span><span class="op">)</span>, </span>
<span>              <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"Z"</span>, <span class="st">"W"</span><span class="op">)</span></span>
<span><span class="va">G</span></span>
<span><span class="co">#&gt;           X   Y Z          W</span></span>
<span><span class="co">#&gt; X 0.0000000 0.5 0  0.3872983</span></span>
<span><span class="co">#&gt; Y 0.0000000 0.0 0  0.3162278</span></span>
<span><span class="co">#&gt; Z 0.4472136 0.0 0 -0.5477226</span></span>
<span><span class="co">#&gt; W 0.0000000 0.0 0  0.0000000</span></span></code></pre></div>
<p>To simulate data from this graph, we can use</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">60000</span>, J <span class="op">=</span> <span class="fl">50000</span>, </span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">50000</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 4 traits.</span></span></code></pre></div>
<p>In the function call above, we specified a GWAS sample size of 60k
for every GWAS with no overlap. We gave different heritabilities of each
of the four traits (the <code>h2</code> argument). Every trait has an
expected 1000 direct effect SNPs (the <code>pi</code> argument). Note
that <code>pi</code> specifies the expected proportion of direct effect
SNPs not the expected number of direct effect SNPs. <code>J</code>
specifies the total number of variants to generate, in this case
50,000.</p>
<div class="section level3">
<h3 id="generating-g-from-xyz-mode-">Generating <span class="math inline">\(G\)</span> from “XYZ”
mode.<a class="anchor" aria-label="anchor" href="#generating-g-from-xyz-mode-"></a>
</h3>
<p>The function <code>xyz_to_G</code> will generate a matrix, <span class="math inline">\(G\)</span>, corresponding to a specific “XYZ” DAG
form. In the “XYZ” DAG, there is an exposure (<span class="math inline">\(X\)</span>), an outcome (<span class="math inline">\(Y\)</span>), and <span class="math inline">\(K-2\)</span> other variables, <span class="math inline">\(Z_1, \dots, Z_{K-2}\)</span>.</p>
<p>There is a (possibly 0) effect of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span> specified by the <code>gamma</code>
argument. Variables <span class="math inline">\(Z_1, \dots,
Z_{K-2}\)</span> can have effects either on or from <span class="math inline">\(X\)</span> and/or <span class="math inline">\(Y\)</span> or both but there are no affects <span class="math inline">\(Z_1, \dots, Z_{K-2}\)</span> on each other.
Effects between each <span class="math inline">\(Z_k\)</span> and <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> respectively are given in the
<code>tau_xz</code> and <code>tau_yz</code> arguments. The direction of
these effects is given in the <code>dir_xz</code> and
<code>dir_yz</code> arguments. All four of these arguments should have
length <span class="math inline">\(k-2\)</span>.</p>
<p>The direction parameters <code>dir_xz</code> and <code>dir_yz</code>
are vectors of 1 or -1 with 1 indicating an effect on <span class="math inline">\(X\)</span> or <span class="math inline">\(Y\)</span> and -1 indicating an effect from <span class="math inline">\(X\)</span> or <span class="math inline">\(Y\)</span>. Effect size arguments
<code>gamma</code>, <code>tau_xz</code>, and <code>tau_yz</code> are
given as signed proportion of variance explained. So if
<code>gamma = -0.3</code>, The direct effect of <span class="math inline">\(X\)</span> explains 30% of the variance of <span class="math inline">\(Y\)</span> and the effect of <span class="math inline">\(X\)</span> on <span class="math inline">\(Y\)</span> is negative.</p>
<p>For example, the code</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/xyz_to_G.html">xyz_to_G</a></span><span class="op">(</span>tau_xz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="op">-</span><span class="fl">0.15</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span><span class="op">)</span>, </span>
<span>                   tau_yz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span>, <span class="op">-</span><span class="fl">0.25</span>, <span class="fl">0.15</span><span class="op">)</span>, </span>
<span>                   dir_xz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>                   dir_yz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                   gamma <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p>generates the matrix corresponding to the graph below:</p>
<div id="htmlwidget-4906484ed42468d86de4" style="width:480px;height:432.632880098888px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-4906484ed42468d86de4">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"Y\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"2,0!\"] \n  \"2\" [label = \"X\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"0,0!\"] \n  \"3\" [label = \"Z1\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"-0.5,1!\"] \n  \"4\" [label = \"Z2\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"1,1!\"] \n  \"5\" [label = \"Z3\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"1,-0.5!\"] \n  \"6\" [label = \"Z4\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fontcolor = \"black\", color = \"black\", fillcolor = \"#FFFFFF\", pos = \"1,-1!\"] \n\"2\"->\"1\" [color = \"black\", label = \"0.548\"] \n\"4\"->\"1\" [color = \"black\", label = \"0.447\"] \n\"6\"->\"1\" [color = \"black\", label = \"0.387\"] \n\"3\"->\"2\" [color = \"black\", label = \"0.316\"] \n\"4\"->\"2\" [color = \"black\", label = \"-0.387\"] \n\"1\"->\"5\" [color = \"black\", label = \"-0.5\"] \n\"2\"->\"5\" [color = \"black\", label = \"0.447\"] \n\"2\"->\"6\" [color = \"black\", label = \"0.548\"] \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p>The weights in the graph give the effect size. Note that this is the
square root of the value provided in <code>tau_xz</code> and
<code>tau_yz</code> which specifies the percent variance explained. For
example, the effect of <span class="math inline">\(Z_1\)</span> on <span class="math inline">\(X\)</span> is <span class="math inline">\(0.316 =
\sqrt{0.1}\)</span> and the effect of <span class="math inline">\(Z_2\)</span> on <span class="math inline">\(X\)</span> is <span class="math inline">\(-0.387 =
- \sqrt{0.15}\)</span>.</p>
</div>
</div>
<div class="section level2">
<h2 id="a-closer-look-at-the-output">A Closer Look at the Output<a class="anchor" aria-label="anchor" href="#a-closer-look-at-the-output"></a>
</h2>
<p>We can now take a look at the output. Summary statistics are
contained in the <code>beta_hat</code> and <code>se_beta_hat</code>
tables which both have dimension <span class="math inline">\(J\times
K\)</span>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "beta_hat"                 "se_beta_hat"             </span></span>
<span><span class="co">#&gt;  [3] "direct_SNP_effects_marg"  "direct_SNP_effects_joint"</span></span>
<span><span class="co">#&gt;  [5] "direct_trait_effects"     "total_trait_effects"     </span></span>
<span><span class="co">#&gt;  [7] "beta_joint"               "beta_marg"               </span></span>
<span><span class="co">#&gt;  [9] "trait_corr"               "R"                       </span></span>
<span><span class="co">#&gt; [11] "R_E"                      "true_h2"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 50000     4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  X             Y             Z             W</span></span>
<span><span class="co">#&gt; [1,]  0.0006032969  0.0074008360  0.0012957544  2.606933e-07</span></span>
<span><span class="co">#&gt; [2,] -0.0023938074 -0.0005043024 -0.0006589290 -3.637417e-03</span></span>
<span><span class="co">#&gt; [3,]  0.0033178233  0.0025144298  0.0080631826  1.817726e-03</span></span>
<span><span class="co">#&gt; [4,]  0.0011088796  0.0060358783  0.0008868393  3.025173e-03</span></span>
<span><span class="co">#&gt; [5,] -0.0074194034 -0.0023891747 -0.0012815257 -1.537479e-03</span></span>
<span><span class="co">#&gt; [6,] -0.0023909927  0.0030201757  0.0061494625  3.285122e-04</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">se_beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]        [,2]        [,3]        [,4]</span></span>
<span><span class="co">#&gt; [1,] 0.004082483 0.004082483 0.004082483 0.004082483</span></span>
<span><span class="co">#&gt; [2,] 0.004082483 0.004082483 0.004082483 0.004082483</span></span>
<span><span class="co">#&gt; [3,] 0.004082483 0.004082483 0.004082483 0.004082483</span></span>
<span><span class="co">#&gt; [4,] 0.004082483 0.004082483 0.004082483 0.004082483</span></span>
<span><span class="co">#&gt; [5,] 0.004082483 0.004082483 0.004082483 0.004082483</span></span>
<span><span class="co">#&gt; [6,] 0.004082483 0.004082483 0.004082483 0.004082483</span></span></code></pre></div>
<p>The <code>direct_trait_effects</code> object is a matrix giving the
input <code>G</code> while <code>total_trait_effects</code> gives the
total effect of each trait on each other trait.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_trait_effects</span></span>
<span><span class="co">#&gt;           X   Y Z          W</span></span>
<span><span class="co">#&gt; X 0.0000000 0.5 0  0.3872983</span></span>
<span><span class="co">#&gt; Y 0.0000000 0.0 0  0.3162278</span></span>
<span><span class="co">#&gt; Z 0.4472136 0.0 0 -0.5477226</span></span>
<span><span class="co">#&gt; W 0.0000000 0.0 0  0.0000000</span></span>
<span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">total_trait_effects</span></span>
<span><span class="co">#&gt;           X         Y Z          W</span></span>
<span><span class="co">#&gt; X 0.0000000 0.5000000 0  0.5454122</span></span>
<span><span class="co">#&gt; Y 0.0000000 0.0000000 0  0.3162278</span></span>
<span><span class="co">#&gt; Z 0.4472136 0.2236068 0 -0.3038068</span></span>
<span><span class="co">#&gt; W 0.0000000 0.0000000 0  0.0000000</span></span></code></pre></div>
<p>True variant effects are also stored in the simulation object. The
<code>direct_SNP_effects_joint</code> and <code>beta_joint</code>
objects give the direct and total SNP effects. The <code>_joint</code>
ending here indicates that these variables store the expected
<em>joint</em> association, conditional on all other variants, i.e. the
causal effects. GWAS effect estimates measure the marginal association,
so we also store the expected direct and total <em>marginal</em>
associations in <code>direct_SNP_effects_marg</code> and
<code>beta_marg</code>. When all variants are independent (no LD), the
expected joint and marginal associations are the same.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_joint</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>However, if we had generated data with LD, this would not be the case
(see below for more on LD). <code>beta_marg</code> is always the
expected value of <code>beta_hat</code>, that is <code>beta_marg</code>
is the association we would measure with infinite GWAS sample size.</p>
<p>Direct SNP effects are always independent across traits while total
SNP effects are the sum of direct effects and indirect effects mediated
by other traits. To see the difference, we make some plots.</p>
<p>First we plot direct SNP effects on <span class="math inline">\(Z\)</span> vs direct SNP effects on <span class="math inline">\(W\)</span></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, </span>
<span>     xlab <span class="op">=</span> <span class="st">"Direct Z effect"</span>, ylab <span class="op">=</span> <span class="st">"Direct W effect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Because direct effects are generated from a point-normal
distribution, most variants have direct effects on at most one of <span class="math inline">\(Z\)</span> or <span class="math inline">\(W\)</span>. However, there are a handful of
variants that have direct effects on both. This is “sporadic pleiotropy”
and can be prevented with <code>sporadic_pleiotropy=FALSE</code> if it
is not desired.</p>
<p>Next we plot the total SNP effects on <span class="math inline">\(Z\)</span> vs the total SNP effects on <span class="math inline">\(W\)</span>. Because <span class="math inline">\(Z\)</span> has a causal effect on <span class="math inline">\(W\)</span>, all variants with effects on <span class="math inline">\(Z\)</span> also affect <span class="math inline">\(W\)</span>. The line in the plot has slope equal
to the total effect of <span class="math inline">\(Z\)</span> on <span class="math inline">\(W\)</span>. The majority SNPs that have non-zero
effect on <span class="math inline">\(Z\)</span> fall exactly on this
line. With <code>sporadic_pleiotropy= FALSE</code>, all of the variants
with non-zero effect on <span class="math inline">\(Z\)</span> would
fall on this line. The variants on the vertical line at 0 are variants
with non-zero direct effect on <span class="math inline">\(W\)</span>
but no direct effect on <span class="math inline">\(Z\)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, </span>
<span>     xlab <span class="op">=</span> <span class="st">"Total Z effect"</span>, ylab <span class="op">=</span> <span class="st">"Total W effect"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">total_trait_effects</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">]</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Finally, we can verify that the observed GWAS effect estimate is a
noisy measurement of <code>beta_marg</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, </span>
<span>     xlab <span class="op">=</span> <span class="st">"beta_marg[,3] = E[beta_hat[,3]]"</span>, ylab <span class="op">=</span> <span class="st">"beta_hat[,3]"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="standardized-vs-non-standardized-effects">Standardized vs Non-Standardized Effects<a class="anchor" aria-label="anchor" href="#standardized-vs-non-standardized-effects"></a>
</h2>
<p>By default, when there is no LD, <code>sim_mv</code> returns
standardized effects and standardized effect estimates. These are the
estimates we would obtain if genotypes were scaled to have population
variance 1 rather than being coded as 0, 1, and 2. The relationship
between standardized and non-standardized effect estimates is simply
<span class="math inline">\(\hat{\beta}_j^{(non-std)} =
\hat{\beta}_j^{(std)}/sd(g_j)\)</span> where <span class="math inline">\(j\)</span> indexes variants, <span class="math inline">\(sd(g_j)\)</span> is the population standard
deviation of variant <span class="math inline">\(j\)</span> (i.e. <span class="math inline">\(\sqrt{2 f_j (1-f_j)}\)</span> where <span class="math inline">\(f_j\)</span> is the allele frequency of variant
<span class="math inline">\(j\)</span>). Similarly the relationship
between the standard error of standardized and non-standardized effect
estimates is <span class="math inline">\(se\left(\hat{\beta}_j^{(non-std)}\right) =
se\left(\hat{\beta}_j^{(std)}\right)/sd(g_j)\)</span>.</p>
<p>In our previously generated <code>sim_dat1</code>, every effect
estimate has the same standard error. This is because the standard error
of <span class="math inline">\(\hat{\beta}_j^{(std)}\)</span> is
approximated as <span class="math inline">\(1/\sqrt{N}\)</span> and we
used a common sample size across all traits. Non-standardized effect
estimates will have different standard errors for different variants
because these also depend on variant allele frequency.</p>
<p>Generating non-standardized effects requires input allele frequencies
which are used to calculate the variance of each SNP. This can be
provided by the <code>af</code> argument which can accept either a
scalar, in which case the same allele frequency is used for every SNP, a
vector of length <code>J</code>, or a function that takes a single
argument and returns a vector of allele frequencies with length
determined by the argument. If the <code>af</code> argument is provided,
<code>sim_mv</code> will return both effect estimates and expected
associations/true effects on the non-standardized scale.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">60000</span>, J <span class="op">=</span> <span class="fl">50000</span>, </span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                  af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 4 traits.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">se_beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]        [,2]        [,3]        [,4]</span></span>
<span><span class="co">#&gt; [1,] 0.016947120 0.016947120 0.016947120 0.016947120</span></span>
<span><span class="co">#&gt; [2,] 0.006562878 0.006562878 0.006562878 0.006562878</span></span>
<span><span class="co">#&gt; [3,] 0.027922349 0.027922349 0.027922349 0.027922349</span></span>
<span><span class="co">#&gt; [4,] 0.008974627 0.008974627 0.008974627 0.008974627</span></span>
<span><span class="co">#&gt; [5,] 0.011949475 0.011949475 0.011949475 0.011949475</span></span>
<span><span class="co">#&gt; [6,] 0.014711105 0.014711105 0.014711105 0.014711105</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simulating-data-with-ld">Simulating Data with LD<a class="anchor" aria-label="anchor" href="#simulating-data-with-ld"></a>
</h2>
<p><code>sim_mv</code> can be used to generate data with LD by inputting
a list of LD matrices and a corresponding table of SNP information. The
function will work fastest if the LD matrix is broken into small blocks.
The input data format for the LD pattern is a list of either a)
matrices, b) sparse matrices (class <code>dsCMatrix</code>) or c) eigen
decompositions (class <code>eigen</code>). The <code>simGWAS</code>
package contains a built in data set containing the LD pattern from
Chromosome 19 in HapMap3 broken into 39 blocks. This LD was estimated
from the HapMap3 European subset using LDShrink. This data set can also
be downloaded <a href="https://zenodo.org/record/6761943#.Yrno2njMIUE" class="external-link">here</a>. The LD
pattern must be accompanied by a data frame containing columns
<code>SNP</code> giving a SNP name (can be arbitrary) and
<code>AF</code> giving the allele frequency.</p>
<p>Let’s look at the built-in LD data</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ld_mat_list"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"snpdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ld_mat_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 39</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ld_mat_list</span>, <span class="va">class</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt;  [7] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [13] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [19] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [25] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [31] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [37] "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This prints the number of SNPs in each block</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ld_mat_list</span>, <span class="va">nrow</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] 140 519 339 435 523 280 675 325 651 548 274 483 442 744 460 177 469 173 358</span></span>
<span><span class="co">#&gt; [20] 564 392 737 596 818 307 863 276 435 204 364 480 381 757 844 753 656 483 856</span></span>
<span><span class="co">#&gt; [39] 709</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ld_mat_list</span>, <span class="va">nrow</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19490</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19490    14</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 14</span></span></span>
<span><span class="co">#&gt;      AF SNP   allele   chr ld_sn…¹     map    pos regio…² snp_id in_ha…³ ldsco…⁴</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 0.394 rs81… G,A       19  7.50<span style="color: #949494;">e</span>7 0   <span style="color: #949494;"> </span>   <span style="text-decoration: underline;">260</span>912       1 7.50<span style="color: #949494;">e</span>7 TRUE       93.0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 0.455 rs81… A,T       19  7.50<span style="color: #949494;">e</span>7 2.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> <span style="text-decoration: underline;">260</span>970       1 7.50<span style="color: #949494;">e</span>7 TRUE       81.4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 0.394 rs81… A,G       19  7.50<span style="color: #949494;">e</span>7 5.59<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> <span style="text-decoration: underline;">261</span>033       1 7.50<span style="color: #949494;">e</span>7 TRUE       93.0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 0.939 rs23… T,C       19  7.50<span style="color: #949494;">e</span>7 2.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="text-decoration: underline;">266</span>034       1 7.50<span style="color: #949494;">e</span>7 TRUE       16.2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 0.399 rs10… C,T       19  7.50<span style="color: #949494;">e</span>7 2.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="text-decoration: underline;">267</span>039       1 7.50<span style="color: #949494;">e</span>7 TRUE       86.8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 0.343 rs12… T,C       19  7.50<span style="color: #949494;">e</span>7 7.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> <span style="text-decoration: underline;">276</span>245       1 7.50<span style="color: #949494;">e</span>7 TRUE       79.3</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 3 more variables: ldscore_hm3 &lt;dbl&gt;, keep_ld_prune_0.1 &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   keep_ld_prune_0.01 &lt;lgl&gt;, and abbreviated variable names ¹​ld_snp_id,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ²​region_id, ³​in_hapmap, ⁴​ldscore_1kg</span></span></span></code></pre></div>
<p>The LD pattern covers 19,490 SNPs. The number of rows in the variant
annotation data frame must match the number of variants in the LD
pattern. If we wish to generate GWAS data for a larger number of
variants than are covered by the LD pattern, <code>simGWAS</code> will
repeat the LD pattern as many times as necessary to give the desired
number of variants.</p>
<p>To generate data with LD, the LD pattern and the variant annotation
data frame are given to the <code>R_LD</code> and <code>snp_info</code>
parameters. The built-in LD pattern corresponds to a density of about
1.2 million variants in per genome. However, in the interest of saving
time and space, we generate data for only 100k variants. This means that
causal effects will be denser than they might be in more realistic
data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sim_dat1_LD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">60000</span>, J <span class="op">=</span> <span class="fl">1e5</span>, </span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">1e5</span>, R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, </span>
<span>                  snp_info <span class="op">=</span> <span class="va">snpdata</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 4 traits.</span></span></code></pre></div>
<p>In data with LD, the <code>_joint</code> objects and
<code>_marg</code> objects are not identical. For example, we can
compare <code>beta_joint</code> and <code>beta_marg</code> for <span class="math inline">\(Z\)</span>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">beta_marg</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span><span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>We can consider variants with non-zero values of
<code>beta_joint[,3]</code> to be variants with causal effects on <span class="math inline">\(Z\)</span> while those with non-zero values of
<code>beta_marg[,3]</code> have non-zero (population) marginal
association with <span class="math inline">\(Z\)</span>, meaning that
they are in LD with at least one causal variant. In the plot, we see
that many variants with no causal effect nevertheless have non-zero
marginal association, as expected. The causal variants don’t fall
exactly on the red line because in some cases, multiple causal variants
are in LD with each other.</p>
<div class="section level3">
<h3 id="ld-pruning-ld-proxies-and-ld-matrix-extraction">LD-Pruning, LD-proxies, and LD matrix extraction<a class="anchor" aria-label="anchor" href="#ld-pruning-ld-proxies-and-ld-matrix-extraction"></a>
</h3>
<p>Many post-GWAS applications such as Mendelian randomization and
polygenic risk score construction require an LD-pruned set of variants.
<code>simGWAS</code> contains a few LD-related functions to help with
pruning and testing methods that require input LD matrices. Note that
all of these methods use the true LD pattern rather than estimated
LD.</p>
<p>The <code>sim_ld_prune</code> function will perform LD-clumping on
simulated data, prioritizing variants according to a supplied
<code>pvalue</code> vector. Although this argument is called
<code>pvalue</code>, it can be any numeric vector used to prioritize
variants. The <code>pvalue</code> argument can also accept an integer.
If <code>pvalue = i</code>, variants will be prioritized according to
the p-value for the trait in the column <code>i</code>. If
<code>pvalue</code> is omitted, variants will be prioritized randomly
(so a different result will be obtained each re-run unless a seed is
set).</p>
<p>To speed up performance, if you only need variants with <span class="math inline">\(p\)</span>-value less than a certain threshold,
supply the <code>pvalue_thresh</code> argument. Below we prune based on
the p-values for trait <span class="math inline">\(Z\)</span> in two
different equivalent ways.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pruned_set1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_ld_prune.html">sim_ld_prune</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">sim_dat1_LD</span>, </span>
<span>                            pvalue <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                            R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, </span>
<span>                            r2_thresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                            pval_thresh <span class="op">=</span> <span class="fl">1e-6</span><span class="op">)</span></span>
<span><span class="co">#&gt; Prioritizing variants based on p-value for trait 3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pruned_set1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 404</span></span>
<span><span class="va">pval3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, <span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">beta_hat</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">/</span><span class="va">se_beta_hat</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pruned_set2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_ld_prune.html">sim_ld_prune</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">sim_dat1_LD</span>, </span>
<span>                            pvalue <span class="op">=</span> <span class="va">pval3</span>, </span>
<span>                            R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, </span>
<span>                            r2_thresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                            pval_thresh <span class="op">=</span> <span class="fl">1e-6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">pruned_set1</span>, <span class="va">pruned_set2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p><code>sim_ld_prune</code> returns a list of indices corresponding to
an LD-pruned set of variants.</p>
<p>The <code>sim_ld_proxy</code> function will return LD-proxies
(variants with LD above a given threshold) with a supplied set of
variants. Here we extract proxies for a few arbitrary variants. The
<code>return_mat</code> option will cause the function to return the LD
matrix for the proxies as well as the indices of proxies</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_proxies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_ld_proxy.html">sim_ld_proxy</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">400</span>, <span class="fl">600</span><span class="op">)</span>, R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, r2_thresh <span class="op">=</span> <span class="fl">0.64</span>, return_mat <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ld_proxies</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [[1]]$index</span></span>
<span><span class="co">#&gt; [1] 100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[1]]$proxy_index</span></span>
<span><span class="co">#&gt; [1] 98 99</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[1]]$Rproxy</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 1.0000000 0.9798682</span></span>
<span><span class="co">#&gt; [2,] 0.9798682 1.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [[2]]$index</span></span>
<span><span class="co">#&gt; [1] 400</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]$proxy_index</span></span>
<span><span class="co">#&gt; [1] 395 396 397 398 399 401 402</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]$Rproxy</span></span>
<span><span class="co">#&gt;           [,1]      [,2]      [,3]      [,4]      [,5]      [,6]      [,7]</span></span>
<span><span class="co">#&gt; [1,] 1.0000000 0.9812294 0.9675572 0.9742288 0.9302869 0.8876271 0.7821419</span></span>
<span><span class="co">#&gt; [2,] 0.9812294 1.0000000 0.9841205 0.9905250 0.9454487 0.9054550 0.7934351</span></span>
<span><span class="co">#&gt; [3,] 0.9675572 0.9841205 1.0000000 0.9810106 0.9367627 0.9175545 0.8097815</span></span>
<span><span class="co">#&gt; [4,] 0.9742288 0.9905250 0.9810106 1.0000000 0.9522425 0.9119614 0.7991366</span></span>
<span><span class="co">#&gt; [5,] 0.9302869 0.9454487 0.9367627 0.9522425 1.0000000 0.8731959 0.7643812</span></span>
<span><span class="co">#&gt; [6,] 0.8876271 0.9054550 0.9175545 0.9119614 0.8731959 1.0000000 0.7522628</span></span>
<span><span class="co">#&gt; [7,] 0.7821419 0.7934351 0.8097815 0.7991366 0.7643812 0.7522628 1.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [[3]]$index</span></span>
<span><span class="co">#&gt; [1] 600</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]$proxy_index</span></span>
<span><span class="co">#&gt; [1] 601 606</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]$Rproxy</span></span>
<span><span class="co">#&gt;            [,1]       [,2]</span></span>
<span><span class="co">#&gt; [1,]  1.0000000 -0.7193148</span></span>
<span><span class="co">#&gt; [2,] -0.7193148  1.0000000</span></span></code></pre></div>
<p>Finally, the <code>sim_extract_ld</code> function will extract the LD
matrix for a set of variants. We can use this to check the previous
functions. First, both of the proxies for index 600 should have
correlation greater than 0.8 with index 600.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_mat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_extract_ld.html">sim_extract_ld</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>, <span class="va">ld_proxies</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">proxy_index</span><span class="op">)</span>, R_LD <span class="op">=</span> <span class="va">ld_mat_list</span><span class="op">)</span></span>
<span><span class="va">ld_mat1</span></span>
<span><span class="co">#&gt;            [,1]       [,2]       [,3]</span></span>
<span><span class="co">#&gt; [1,]  1.0000000 -0.8044757  0.8670863</span></span>
<span><span class="co">#&gt; [2,] -0.8044757  1.0000000 -0.7193148</span></span>
<span><span class="co">#&gt; [3,]  0.8670863 -0.7193148  1.0000000</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">ld_mat1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
<p>Next we can check that the LD-pruned set have no mutual correlations
greater than <span class="math inline">\(\sqrt(0.1) \approx
0.316\)</span></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_mat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_extract_ld.html">sim_extract_ld</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, index <span class="op">=</span> <span class="va">pruned_set1</span>, R_LD <span class="op">=</span> <span class="va">ld_mat_list</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">ld_mat2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span> <span class="co"># remove the diagonal</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">ld_mat2</span><span class="op">^</span><span class="fl">2</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="sample-overlap">Sample Overlap<a class="anchor" aria-label="anchor" href="#sample-overlap"></a>
</h2>
<p>If two GWAS are performed on different traits using overlapping
samples, the sampling errors of effect estimates will be correlated. If
the two GWAS have sample sizes <span class="math inline">\(N_1\)</span>
and <span class="math inline">\(N_2\)</span> with <span class="math inline">\(N_c\)</span> overlapping samples, then the
correlation of <span class="math inline">\(\hat{z}_{1j}\)</span> and
<span class="math inline">\(\hat{z}_{2j}\)</span>, <span class="math inline">\(z\)</span>-scores for variant <span class="math inline">\(j\)</span> in study 1 and study 2, is
approximately <span class="math inline">\(\frac{N_c}{\sqrt{N_1 N_2}}
\rho_{1,2}\)</span> where <span class="math inline">\(\rho_{1,2}\)</span> is the trait correlation.</p>
<p>In <code>sim_mv</code> it is possible to specify sample overlap
between studies by providing <code>N</code> as a matrix rather than a
scalar or a vector. If <code>N</code> is a matrix than
<code>N[i,i]</code> is the sample size of study <span class="math inline">\(i\)</span> and <code>N[i,j]</code> is the number
of overlapping samples between studies <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>. In the model used by
<code>simGWAS</code>, all traits can be decomposed into a genetic
component and an environmental component. We always assume the
environmental trait component is independent of the genetic trait
component. Correlation between the genetic components of the specified
traits is determined by the supplied DAG. It is also possible to supply
the correlation between the environmental components of the traits using
the <code>R_E</code> parameter.</p>
<p>Below is an example using the same 4 trait DAG we have been using so
far. Below we specify that all four GWAS have sample size 60,000 and
there are 30,000 samples overlapping between any pair of GWAS. Note that
<code>simGWAS</code> assumes that the trait correlation is the same in
any subset of samples so it doesn’t matter if the same 30,000 samples
are in all studies or if different sets overlap between different pairs
of studies. We also specify an environmental correlation matrix in which
the pairwise correlation in environmental effects for any pair of traits
is 0.4. <code>R_E</code> defaults to the identity if it is not
provided.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">R_E</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.4</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">R_E</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">30000</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">N</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">60000</span></span>
<span><span class="va">N</span></span>
<span><span class="co">#&gt;       [,1]  [,2]  [,3]  [,4]</span></span>
<span><span class="co">#&gt; [1,] 60000 30000 30000 30000</span></span>
<span><span class="co">#&gt; [2,] 30000 60000 30000 30000</span></span>
<span><span class="co">#&gt; [3,] 30000 30000 60000 30000</span></span>
<span><span class="co">#&gt; [4,] 30000 30000 30000 60000</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sim_dat3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                  N <span class="op">=</span> <span class="va">N</span>, J <span class="op">=</span> <span class="fl">1e5</span>, R_E <span class="op">=</span> <span class="va">R_E</span>,</span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">1e5</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 4 traits.</span></span></code></pre></div>
<p>The simulation data object contains three matrices that describe
between study correlation. <code>R_E</code> is the environmental
correlation supplied in the command option, <code>trait_corr</code> is
the total trait correlation, and <code>R</code> is the correlation
between noise terms for pairs of GWAS which is obtained by scaling
<code>trait_corr</code> by the appropriate values.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat3</span><span class="op">$</span><span class="va">R_E</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4]</span></span>
<span><span class="co">#&gt; [1,]  1.0  0.4  0.4  0.4</span></span>
<span><span class="co">#&gt; [2,]  0.4  1.0  0.4  0.4</span></span>
<span><span class="co">#&gt; [3,]  0.4  0.4  1.0  0.4</span></span>
<span><span class="co">#&gt; [4,]  0.4  0.4  0.4  1.0</span></span>
<span><span class="va">sim_dat3</span><span class="op">$</span><span class="va">trait_corr</span></span>
<span><span class="co">#&gt;           X         Y          Z          W</span></span>
<span><span class="co">#&gt; X 1.0000000 0.4300000 0.46024999 0.30037881</span></span>
<span><span class="co">#&gt; Y 0.4300000 1.0000000 0.34844659 0.35095546</span></span>
<span><span class="co">#&gt; Z 0.4602500 0.3484466 1.00000000 0.06718562</span></span>
<span><span class="co">#&gt; W 0.3003788 0.3509555 0.06718562 1.00000000</span></span>
<span><span class="va">sim_dat3</span><span class="op">$</span><span class="va">R</span></span>
<span><span class="co">#&gt;           X         Y          Z          W</span></span>
<span><span class="co">#&gt; X 1.0000000 0.2150000 0.23012499 0.15018940</span></span>
<span><span class="co">#&gt; Y 0.2150000 1.0000000 0.17422330 0.17547773</span></span>
<span><span class="co">#&gt; Z 0.2301250 0.1742233 1.00000000 0.03359281</span></span>
<span><span class="co">#&gt; W 0.1501894 0.1754777 0.03359281 1.00000000</span></span></code></pre></div>
<p>If there is no sample overlap, the <code>R_E</code> argument to
<code>sim_mv</code> will have no affect on summary statistics but will
influence the calculation of <code>trait_corr</code>. For example, in
our previous data generated without sample overlap and in which we did
not specify <code>R_E</code> we have</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat1_LD</span><span class="op">$</span><span class="va">R_E</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4]</span></span>
<span><span class="co">#&gt; [1,]    1    0    0    0</span></span>
<span><span class="co">#&gt; [2,]    0    1    0    0</span></span>
<span><span class="co">#&gt; [3,]    0    0    1    0</span></span>
<span><span class="co">#&gt; [4,]    0    0    0    1</span></span>
<span><span class="va">sim_dat1_LD</span><span class="op">$</span><span class="va">trait_corr</span></span>
<span><span class="co">#&gt;            X          Y          Z           W</span></span>
<span><span class="co">#&gt; X 1.00000000 0.15000000  0.2236068  0.04114918</span></span>
<span><span class="co">#&gt; Y 0.15000000 1.00000000  0.1118034  0.09172584</span></span>
<span><span class="co">#&gt; Z 0.22360680 0.11180340  1.0000000 -0.15190340</span></span>
<span><span class="co">#&gt; W 0.04114918 0.09172584 -0.1519034  1.00000000</span></span>
<span><span class="va">sim_dat1_LD</span><span class="op">$</span><span class="va">R</span></span>
<span><span class="co">#&gt;   X Y Z W</span></span>
<span><span class="co">#&gt; X 1 0 0 0</span></span>
<span><span class="co">#&gt; Y 0 1 0 0</span></span>
<span><span class="co">#&gt; Z 0 0 1 0</span></span>
<span><span class="co">#&gt; W 0 0 0 1</span></span></code></pre></div>
<p>The <code>R</code> matrix will always be the identity if there is no
sample overlap.</p>
</div>
<div class="section level2">
<h2 id="generating-data-for-multiple-gwas-with-the-same-variant-effects">Generating Data for Multiple GWAS with the Same Variant Effects<a class="anchor" aria-label="anchor" href="#generating-data-for-multiple-gwas-with-the-same-variant-effects"></a>
</h2>
<p>In some cases, it may be desirable to generate multiple sets of
summary statistics for the same set of effect sizes. This would mimic
performing multiple GWAS on the same trait. The function
<code>gen_bhat_from_b</code> allows this. This function takes as input a
matrix of either standardized or non-standardized joint (causal) effects
to <code>b_joint_std</code> or <code>b_joint</code>. The function also
requires the trait correlation matrix and the sample size argument.
Optionally, LD arguments <code>R_LD</code> and <code>snp_info</code> can
be supplied, or the allele frequency can be supplied via
<code>af</code>. Most users will not need the other options.</p>
<p>Below, we generate new GWAS data for the effects in
<code>sim_dat1_LD</code>. Here we use the same LD pattern used
originally but this is not required. For example, we could simulate GWAS
in samples with different ancestry but supplying a different LD
pattern.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat2_LD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gen_bhat_from_b.html">gen_bhat_from_b</a></span><span class="op">(</span>b_joint <span class="op">=</span> <span class="va">sim_dat1_LD</span><span class="op">$</span><span class="va">beta_joint</span>, </span>
<span>                               trait_corr <span class="op">=</span> <span class="va">sim_dat1_LD</span><span class="op">$</span><span class="va">trait_corr</span>, </span>
<span>                               N <span class="op">=</span> <span class="fl">40000</span>, </span>
<span>                               R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, snp_info <span class="op">=</span> <span class="va">snpdata</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 4 traits.</span></span></code></pre></div>
<p>The resulting simulation data object only contains effect estimates,
standard errors, <code>R</code>, and <code>snp_info</code>. All other
information should match the original data set.</p>
<p>Below we compare effect estimates in the new study and the old study
for trait <span class="math inline">\(X\)</span>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat2_LD</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">sim_dat1_LD</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>Our first simulation object <code>sim_dat1</code> was generated with
no LD and without supplying the <code>af</code> argument, so the
<code>beta_joint</code> values for that object represent standardized
effects. If we were generating secondary studies from this object, we
should use the <code>b_joint_std</code> argument rather than
<code>b_joint</code> in the call to <code>get_bhat_from_b</code>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
