<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Simulate multivariate GWAS data"><title>Simulate multivariate GWAS data — sim_mv • simGWAS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate multivariate GWAS data — sim_mv"><meta property="og:description" content="Simulate multivariate GWAS data"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">simGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0.0037</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/simGWAS/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulate multivariate GWAS data</h1>
      <small class="dont-index">Source: <a href="https://github.com/jean997/simGWAS/blob/HEAD/R/simulate2.R" class="external-link"><code>R/simulate2.R</code></a></small>
      <div class="d-none name"><code>sim_mv.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Simulate multivariate GWAS data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sim_mv</span><span class="op">(</span></span>
<span>  <span class="va">N</span>,</span>
<span>  <span class="va">J</span>,</span>
<span>  <span class="va">h2</span>,</span>
<span>  <span class="va">pi</span>,</span>
<span>  <span class="va">G</span>,</span>
<span>  R_E <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_LD <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  snp_info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  af <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sporadic_pleiotropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_dat <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>N</dt>
<dd><p>GWAS sample size. N can be a scalar, vector, or matrix. If N is a scalar, all GWAS have the same sample size
and there is no overlap between studies. If N is a vector, each element of N specifies the sample size of the corresponding
GWAS and there is no overlap between studies. If N is a matrix, N_ii specifies the sample size of study i
and N_ij specifies the number of samples present in both study i and study j. The elements of N must be positive
but non-integer values will not generate an error.</p></dd>


<dt>J</dt>
<dd><p>Number of variants to simulate</p></dd>


<dt>h2</dt>
<dd><p>A scalar or vector giving the heritability of each trait.</p></dd>


<dt>pi</dt>
<dd><p>A scalar or vector giving the expected proportion of direct effect SNPs for each trait.</p></dd>


<dt>G</dt>
<dd><p>Matrix of direct effects. Rows correspond to the 'from' trait
and columns correspond to the 'to' trait, so G[1,2] is the direct effect of trait 1 on trait 2. G should have 0
on the diagonal.</p></dd>


<dt>R_E</dt>
<dd><p>Environmental correlation between traits. R_E is ignored if there is no sample overlap.</p></dd>


<dt>R_LD</dt>
<dd><p>List of eigen decompositions of LD correlation matrices, may be missing.</p></dd>


<dt>snp_info</dt>
<dd><p>If R_LD is provided, provide a data frame with columns "SNP" and "AF"</p></dd>


<dt>af</dt>
<dd><p>Optional vector of allele frequencies.
This option is only used when data are generated without LD.
Otherwise the allele frequency in snp_info is used instead.</p></dd>


<dt>sporadic_pleiotropy</dt>
<dd><p>Allow sporadic pleiotropy between traits. Defaults to TRUE.</p></dd>


<dt>return_dat</dt>
<dd><p>Useful for debugging.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list with the following elements:</p>


<p>Simulated effect estimates and standard errors are contained in matrices</p>


<p>+ <code>beta_hat</code>: Effect estimates for each trait</p>


<p>+ <code>se_beta_hat</code> Standard error of effect estimates</p>


<p>Everything else returned has to do with simulation parameters or true effects</p>


<p>+ <code>direct_SNP_effects_marg</code> and <code>direct_SNP_effects_joint</code> give direct effects of SNPs on traits. These are the
same if there is no LD. If there is LD, <code>direct_SNP_effects_marg</code> is the direct component of the expected marginal association.</p>


<p>+ <code>beta_marg</code> and <code>beta_joint</code> give the total SNP effects (direct and indirect). These are the
same if there is no LD. If there is LD, <code>beta_marg</code> is the total expected marginal association. i.e. <code>beta_marg</code> is the
expected value of <code>beta_hat</code>.</p>


<p>+ <code>direct_trait_effects</code> and <code>total_trait_effects</code> matrices giving direct and total effects of traits on each other.
<code>direct_trait_effects</code> should be equal to the supplied <code>G</code>.</p>


<p>+ <code>R</code> gives the row correlation of <code>beta_hat - beta_marg</code>.</p>


<p>Most users can ignore everything else returned.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function generates GWAS summary statistics from a linear SEM specified by the matrix G.
The previous "xyz" mode is now deprecated. If you are used to using this function in xyz mode,
you can generate the corresponding G using the xyz_to_G function (see examples).</p>
<p>G should be square nxn matrix with 0s on the diagonal. <code>h2</code> and <code>pi</code> should be scalars or have length n.
All variables have variance 1, so <code>G[i,j]^2</code> is the proportion of variance of trait j explained by the effect of trait i.</p>
<p>You should get an error if you specify a cyclic DAG, a DAG that is impossible, or a combination of <code>G</code> and <code>h2</code>
that is impossible.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Two traits with no causal relationship and some environmental correlation</span></span></span>
<span class="r-in"><span><span class="co"># specify completely overlapping GWAS</span></span></span>
<span class="r-in"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1000</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">R_E</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.8</span>, <span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">sim_mv</span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>, J <span class="op">=</span> <span class="fl">20000</span>, h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.3</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">20000</span>,</span></span>
<span class="r-in"><span>               G <span class="op">=</span> <span class="va">G</span>, R_E <span class="op">=</span> <span class="va">R_E</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20000 SNPs and 2 traits.</span>
<span class="r-in"><span><span class="va">dat</span><span class="op">$</span><span class="va">R</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [,1]      [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 1.0000000 0.5184593</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.5184593 1.0000000</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">beta_hat</span> <span class="op">-</span> <span class="va">dat</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [,1]      [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 1.0000000 0.5201858</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.5201858 1.0000000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Use xyz_to_G to generate G from xyz specification</span></span></span>
<span class="r-in"><span><span class="va">myG</span> <span class="op">&lt;-</span> <span class="fu"><a href="xyz_to_G.html">xyz_to_G</a></span><span class="op">(</span>tau_xz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="op">-</span><span class="fl">0.3</span><span class="op">)</span>, tau_yz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.25</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        dir_xz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, dir_yz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, gamma <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># If N is a scalar or a vector, there is no sample overlap</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">sim_mv</span><span class="op">(</span>N <span class="op">=</span> <span class="fl">10000</span>, J <span class="op">=</span> <span class="fl">20000</span>, h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">4</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span>, <span class="fl">1000</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">/</span><span class="fl">20000</span>,</span></span>
<span class="r-in"><span>              G <span class="op">=</span> <span class="va">myG</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20000 SNPs and 4 traits.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">dat</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">dat</span><span class="op">$</span><span class="va">total_trait_effects</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="sim_mv-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

